### ISSUE TITLE
Validation for HPA and Deployment/replicas

### ISSUE THREAD

**Issue by shaoxt** [argoproj/argo-cd#1603]
**Is your feature request related to a problem? Please describe.**
The suggested way to apply HPA on Deployment is to get rid of "replicas" in DeploymentSpec.
But it scales down to 1 pod when deleting the replicas on existing Deployment.
Here is the issue
https://github.com/kubernetes/kubernetes/issues/25238

**Describe the solution you'd like**

## argocd app **hpa** sample-namespace

1.  Check whether there is any HorizontalPodAutoscaler
    - If YES, check whether the target is a **Deployment**
        - If YES, goto step 2
        - If NO, return
    - If NO,  return
2. Check whether the Deployment has **replicas**
    - If YES, show error, suggest to delete the replicas
    - If NO, goto step 3
3. Sync HorizontalPodAutoscaler **ONLY** to make sure it get synced.
    - If SUCCESS, goto step 4
    - If NO, show error
4. Check whether the old Deployment has **replicas**
   - If YES, run "kubectl apply set-last-applied -f deployment.yaml -n ${THE_NAMESPACE}", deployment.yaml  is the **OLD** Deployment Spec
   - If NO, return

## argocd app sync  sample-namespace

HPA Sample:
```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  annotations:
  labels:
    app: hpa-samples
  name: hpa-samples
spec:
  minReplicas: 3
  maxReplicas: 30
  metrics:
  scaleTargetRef:
    apiVersion: apps/v1beta2
    kind: Deployment
    name: hpa-samples
```

**Describe alternatives you've considered**

**Additional context**

**Comment by edlee2121** [comment]
I think this is a feature that would be useful for anyone using HPA. This reminds me of how argocd currently uses `auth reconcile` rather than `apply` when updating roles and rolebindings.

**Comment by alexec** [comment]
I think we may be planning this @alexmt ?

**Comment by soerenmartius** [comment]
I am currently having exactly the same issue. Since this issue has been closed - is there a best practice on how to deal with it?

**Comment by beyondbill** [comment]
Although the issue has been marked as completed, I do not think it's actually fixed. Here are what I did in sequence for testing but still lead the Deployment's replicas to be set to 1

- Create a Deployment with replicas = 2 via ArgoCD
- Create a HPA targeting at the Deployment with minReplicas = 3 via ArgoCD. The HPA sets Deployment's replicas = 3, as expected
- Disable ArgoCD autoSync and remove the field replicas from Git, as seemingly suggested by the official [doc](https://argo-cd.readthedocs.io/en/stable/user-guide/best_practices/#leaving-room-for-imperativeness). I expect ArgoCD to ignore the field replicas in the Deployment. However, ArgoCD still shows a diff that Deployment's replicas should be 1 (default value when left unset)
- Click sync on ArgoCD, and the Deployment is downsized to only 1 pod, as expected from the previous step
- After ~1 min, HPA scales the Deployment back to 3 replicas

How could I avoid the transient phase of downsizing to 1 pod when adding a HPA to a Deployment?

**Comment by Somefive** [comment]
@beyondbill We also encounter this issue and found the application-level diff ignorence could solve this issue. This requires extra configuration for the ArgoCD application. https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#application-level-configuration

**Comment by beyondbill** [comment]
@Somefive I did the exact same as the linked doc suggests. Argo did ignore the diff on replicas when replicas field was removed. However, when image field was changed in the same Deployment template, Argo applied both the replicas and the image changes all together. I consider this as a bug rather than a feature. When ignoreDiff is deliberately set, people want Argo to leave it totally outside Argo management.
